splash/browser_tab.py

```diff
@@ -768,3 +838,146 @@ def log(self, message, min_level=None):
 
         message = "[%s] %s" % (self.uid, message)
         log.msg(message, system='render')
+
+
+class OneShotCallbackProxy(QObject):
+    """
+    A proxy object that allows JavaScript to run Python callbacks.
+
+    This creates a JavaScript-compatible object (can be added to `window`)
+    that has functions `resume()` and `error()` that can be connected to
+    Python callbacks.
+
+    It is "one shot" because either `resume()` or `error()` should be called
+    exactly _once_. It raises an exception if the combined number of calls
+    to these methods is greater than 1.
+
+    If timeout is zero, then the timeout is disabled.
+    """
+
+    def __init__(self, parent, callback, errback, timeout=0):
+        self.name = str(uuid.uuid1())
+        self._used_up = False
+        self._callback = callback
+        self._errback = errback
+
+        if timeout < 0:
+            raise ValueError('OneShotCallbackProxy timeout must be >= 0.')
+        elif timeout == 0:
+            self._timer = None
+        elif timeout > 0:
+            self._timer = QTimer()
+            self._timer.setSingleShot(True)
+            self._timer.timeout.connect(self._timed_out)
+            self._timer.start(timeout * 1000)
+
+        super(OneShotCallbackProxy, self).__init__(parent)
+
+    @pyqtSlot()
+    @pyqtSlot('bool')
+    @pyqtSlot('double')
+    @pyqtSlot('QByteArray')
+    @pyqtSlot('QVariantMap')
+    def resume(self, value=None):
+        if self._used_up:
+            raise OneShotCallbackError("resume() called on a one shot" \
+                                       " callback that was already used up.")
+
+        self._use_up()
+        self._callback(self._qt_object_to_py_object(value))
+
+    @pyqtSlot(str)
+    def error(self, message):
+        if self._used_up:
+            raise OneShotCallbackError("error() called on a one shot" \
+                                       " callback that was already used up.")
+        self._use_up()
+        self._errback(str(message))
+
+    def cancel(self, reason):
+        self._use_up()
+        self._errback("One shot callback canceled due to: %s." % reason)
+
+    def _dict_looks_like_list(self, dict_):
+        """
+        QT converts JS arrays and objects into QVariantMap, which it then
+        converts to Python dict.
+
+        The problem with this behavior is that we don't know if the dict
+        we received started life as a JS Array or a generic JS object. We
+        use a simple heuristic to determine if dict_ "looks like" a
+        list: the dict_ is a list if dict_.keys() contains all of the integers
+        in the range [0,n-1], where n is the number of keys in the dict_.
+        """
```

Comment:
Agreed. It's also ambiguous for an empty array/object. If we explicitly disallow arrays, that actually simplifies a lot of the other code surrounding this. I will clean it up and update the docs.